sdkit/filter/codeformer.py

```diff
@@ -0,0 +1,240 @@
+import sys
+import os
+import cv2
+import torch
+import torch.nn.functional as F
+import argparse
+
+from sdkit import Context
+
+from torchvision.transforms.functional import normalize
+
+from basicsr.archs.rrdbnet_arch import RRDBNet
+from basicsr.utils import imwrite, img2tensor, tensor2img
+from basicsr.utils.download_util import load_file_from_url
+from sdkit.modules.codeformer.utils.misc import gpu_is_available, get_device
+from sdkit.modules.codeformer.utils.realesrgan_utils import RealESRGANer
+from sdkit.modules.codeformer.utils.registry import ARCH_REGISTRY
+from sdkit.modules.codeformer.utils.logger import get_root_logger
+
+from sdkit.modules.codeformer.archs.codeformer_arch import CodeFormer
+from sdkit.modules.codeformer.facelib.utils.face_restoration_helper import FaceRestoreHelper
+from sdkit.modules.codeformer.facelib.utils.misc import is_gray
+
+import traceback
+import warnings
+import sys
+
+"""
+def warn_with_traceback(message, category, filename, lineno, file=None, line=None):
+
+    log = file if hasattr(file,'write') else sys.stderr
+    traceback.print_stack(file=log)
+    log.write(warnings.formatwarning(message, category, filename, lineno, line))
+
+warnings.showwarning = warn_with_traceback
+"""
+
+def CodeFormer_Initialize():
+    pretrain_model_url = {
+        'codeformer': 'https://github.com/sczhou/CodeFormer/releases/download/v0.1.0/codeformer.pth',
+        'detection': 'https://github.com/sczhou/CodeFormer/releases/download/v0.1.0/detection_Resnet50_Final.pth',
+        'parsing': 'https://github.com/sczhou/CodeFormer/releases/download/v0.1.0/parsing_parsenet.pth',
+        'realesrgan': 'https://github.com/sczhou/CodeFormer/releases/download/v0.1.0/RealESRGAN_x2plus.pth'
+    }
+
+    # download weights
+    for model_type in pretrain_model_url:
+        if not os.path.exists(f'CodeFormer/weights/{model_type}/{model_type}.pth'):
+            load_file_from_url(url=pretrain_model_url[model_type], model_dir=f'CodeFormer/weights/{model_type}', progress=True, file_name=None)
```

Comment:
This would create the `Codeformer/weights` folders in the `.../stable-diffusion/` folder, and not in the `.../models/` hierarchy.

Reply:
Any recommendation of how/where to get the model root path from? (e.g. C:\EasyDiffusion\Models)

Reply:
No, unfortunately, not. I hoped someone else might :-)